### PR TITLE
L110: C# nullable reference types proposal

### PR DESCRIPTION
**Pull request by tonydnewell**
Proposal to support nullable reference types in the protocol buffers and gRPC plugin code generation.

### CODE REVIEW

File: L107-csharp-nullable-reference-types.md [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+- [Investigate support for nullable reference types in generated code](https://github.com/grpc/grpc-dotnet/issues/606)
+- [Add option to generate C# nullable annotations](https://github.com/protocolbuffers/protobuf/issues/6632)
+- [Support C# 8 nullable type references in Grpc.Tools](https://github.com/grpc/grpc/issues/20729)
+
+Currently the code that is generated is compatible with C# 7.3 and is not in a 
+*nullable context* since the `// <auto-generated>` comment in the code disables it.
+
+No nullable checks are made and no compiler warnings are generated for this code,
+even when the code is compiled in a project with nullable checks enabled.
+This  means that some nullable checks that could be useful in user code are not done.
```
Comment (jskeet) [inline L107-csharp-nullable-reference-types.md:38]: Mega nit: double space after "This"

File: L107-csharp-nullable-reference-types.md [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+
+**Note:** here we are *not* proposing any changes to the API or semantics to support protobuf optional fields as nullable, as sometimes requested (e.g. - [C#: Consider exposing protobuf optional fields as nullable properties](https://github.com/protocolbuffers/protobuf/issues/9083)). That has already been addressed in the above issue.
```
Comment (jskeet) [inline L107-csharp-nullable-reference-types.md:40]: Possibly unwrap this, to be consistent with other paragraphs?

**Review by jskeet — APPROVED** [review]
Looks good. I think it would be worth mentioning the Google.Protobuf runtime, and how it doesn't currently have any nullability annotations. There's work ongoing to implement that, but it's very slow-moving (partly due to my lack of reviewing time).

**Review by JamesNK — APPROVED** [review]

File: L107-csharp-nullable-reference-types.md [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+
+Add the `enable_nrt` option, and when the enabled:
+- add `#nullable` directive
+- add annotations to Message fields as they can be null
+- `Equals(obj)` and `Equals(T)` would have nullable parameters
+- `MergeFrom(T)` would have a nullable parameter
+
+### gRPC plugin code generation
+
+Add the `enable_nrt` option, and when the enabled:
```
Comment (jtattermusch) [inline L107-csharp-nullable-reference-types.md:145]: typo: "and when the enabled"

File: L107-csharp-nullable-reference-types.md [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+  - https://github.com/grpc/grpc/tree/master/src/csharp/Grpc.Tools 
+
+It will require collaboration between the Protocol Buffers team and the gRPC team to
+coordinate the release of this feature.
+
+### Protocol buffers code generation
+
+Add the `enable_nrt` option, and when the enabled:
+- add `#nullable` directive
+- add annotations to Message fields as they can be null
```
Comment (jtattermusch) [inline L107-csharp-nullable-reference-types.md:139]: I didn't know that we know so precisely what exactly what would change (sounded like the reality will be more complicated that this), but if we do, that's great. 
FWIW, I think we don't need to mention exactly what changes would be required here, as long as we describe the general idea behind the changes and approx. what they would be. So I think we could at least qualify the statements here to not be as prescriptive?

File: L107-csharp-nullable-reference-types.md [hunk 5]
```diff
@@ -0,0 +1,169 @@
...
+- gprc/grpc PR 33878 [Add null reference types support to gRPC C# plugin](https://github.com/grpc/grpc/pull/33878)
+- grpc/grpc PR 33879 [Grpc.Tools changes to support nullable reference types](https://github.com/grpc/grpc/pull/33879)
+
+## Related work
+There is separate and ongoing work to implement nullable annotations in the
+[Google.Protobuf](https://github.com/protocolbuffers/protobuf/tree/main/csharp/src/Google.Protobuf)
+runtime. This is not part of this proposal which is focused on the code generation.
+
+See:
+- [#10170 Set all but main project to nullable ](https://github.com/protocolbuffers/protobuf/pull/10170)
```
Comment (jtattermusch) [inline L107-csharp-nullable-reference-types.md:169]: nit: remove space before `]'

**Review by jtattermusch — COMMENTED** [review]
The proposal looks good to me (a few comments pending), but I'll leave it to @apolcyn, @jskeet  and @JamesNK  to actually LGTM.

@apolcyn will be listed as the Approver in the top of this grpc-proposal (as clearly proposals need to be approved by someone in the grpc team).

File: L107-csharp-nullable-reference-types.md (hunk 4 above)
Comment (tonydnewell) [reply L107-csharp-nullable-reference-types.md:139]: I've now qualified this by saying that other changes may become apparent during development.

**Review by apolcyn — APPROVED** [review]

File: L107-csharp-nullable-reference-types.md [hunk 6]
```diff
@@ -0,0 +1,172 @@
+Handling of null reference types in C# code generation
```
Comment (apolcyn) [inline L107-csharp-nullable-reference-types.md:1]: @tonydnewell can you please update this file name to `L110-csharp-nullable-reference-types.md`.

We can merge after that.

**Review by apolcyn — APPROVED** [review]

File: L107-csharp-nullable-reference-types.md (hunk 6 above)
Comment (tonydnewell) [reply L107-csharp-nullable-reference-types.md]: @apolcyn renamed.

**Pull request merged**